File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/FieldConverter.scala
package com.exasol.cloudetl.kafka.deserialization

import java.lang.{Long => JLong}
import java.sql.Timestamp

class FieldConverter(outputColumnTypes: Seq[Class[_]]) {

  final def convertRow(row: Seq[Any]): Seq[Any] =
    row.zipWithIndex.map(x => convert(outputColumnTypes(x._2), x._1))

  final def convert(columnType: Class[_], value: Any): Any =
    value match {
      case x: JLong if columnType == classOf[Timestamp] => new Timestamp(x)
      case _                                            => value
    }

}

File: src/main/java/com/exasol/cloudetl/kafka/deserialization/FieldConverter.java
package com.exasol.cloudetl.kafka.deserialization;

import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;

public class FieldConverter {
    private final List<Class<?>> outputColumnTypes;

    public FieldConverter(final List<Class<?>> outputColumnTypes) {
        this.outputColumnTypes = outputColumnTypes;
    }

    public List<Object> convertRow(final List<Object> row) {
        final List<Object> rowValues = row;
        final List<Class<?>> columnTypes = this.outputColumnTypes;
        final List<Object> result = new ArrayList<>();
        for (int index = 0; index < rowValues.size(); index++) {
            result.add(convert(columnTypes.get(index), rowValues.get(index)));
        }
        return result;
    }

    public Object convert(final Class<?> columnType, final Object value) {
        if ((value instanceof Long) && (columnType == Timestamp.class)) {
            return new Timestamp((Long) value);
        }
        return value;
    }
}
